src/utils/tcti.rs

```diff
@@ -0,0 +1,451 @@
+    assert_eq!(
+        tcti,
+        Tcti::Device(DeviceConfig {
+            path: Some(PathBuf::from("/try/this/path")),
+        })
+    );
+
+    let tcti = Tcti::from_str("device").unwrap();
+    assert_eq!(tcti, Tcti::Device(DeviceConfig { path: None }));
+
+    let tcti = Tcti::from_str("tabrmd:bus_name=some.bus.Name2,bus_type=session").unwrap();
+    assert_eq!(
+        tcti,
+        Tcti::Tabrmd(TabrmdConfig {
+            bus_name: Some(String::from("some.bus.Name2")),
+            bus_type: Some(BusType::Session)
+        })
+    );
+
+    let tcti = Tcti::from_str("tabrmd").unwrap();
+    assert_eq!(
+        tcti,
+        Tcti::Tabrmd(TabrmdConfig {
+            bus_name: None,
+            bus_type: None
+        })
+    );
+}
+
+/// Configuration for a Device TCTI context
+///
+/// The default configuration uses the library default of
+/// `/dev/tpm0`.
+#[derive(Clone, Debug, Default, PartialEq)]
+pub struct DeviceConfig {
+    /// Path to the device node to connect to
+    ///
+    /// If set to `None`, the default location is used
+    path: Option<PathBuf>,
+}
+
+impl FromStr for DeviceConfig {
+    type Err = Error;
+
+    fn from_str(config_str: &str) -> Result<Self, Self::Err> {
+        if config_str.is_empty() {
+            return Ok(Default::default());
+        }
+
+        Ok(DeviceConfig {
+            path: Some(PathBuf::from(config_str)),
+        })
+    }
+}
+
+#[test]
+fn validate_from_str_device_config() {
+    let config = DeviceConfig::from_str("").unwrap();
+    assert!(config.path.is_none());
+
+    let config = DeviceConfig::from_str("/dev/tpm0").unwrap();
+    assert_eq!(config.path.unwrap(), PathBuf::from("/dev/tpm0"));
+}
+
+/// Configuration for an Mssim TCTI context
+///
+/// The default configuration will point to `localhost:2321`
+#[derive(Clone, Debug, Default, PartialEq)]
+pub struct MssimConfig {
+    /// Address of the server to connect to
+    ///
+    /// If set to `None`, `localhost` is used
+    host: Option<ServerAddress>,
+    /// Port used by the server at the address given in `host`
+    ///
+    /// If set to `None`, `2321` is used
+    port: Option<u16>,
+}
+
+impl FromStr for MssimConfig {
+    type Err = Error;
+
+    fn from_str(config_str: &str) -> Result<Self, Self::Err> {
+        if config_str.is_empty() {
+            return Ok(Default::default());
+        }
+        let host_pattern = Regex::new(r"(,|^)host=(.*?)(,|$)").unwrap(); // should not fail
+        let host = match host_pattern.captures(config_str) {
+            None => None,
+            Some(captures) => Some(ServerAddress::from_str(captures.get(2).unwrap().as_str())?),
+        };
+
+        let port_pattern = Regex::new(r"(,|^)port=(.*?)(,|$)").unwrap(); // should not fail
+        let port = match port_pattern.captures(config_str) {
+            None => None,
+            Some(captures) => Some(
+                u16::from_str(captures.get(2).unwrap().as_str())
```

Comment:
Is the `unwrap` always ok?

Reply:
No, actually it isn't - I updated it locally but didn't feel like pushing again. It's essentially the same fix as for the first comment you posted. I'll add a test with something like `host=,port=1234`, that would've caught it.

Reply:
(I also updated all the other places that do it like this)